File: neuralnetwork/jsonReader.cpp
#include <fstream>
#include <iostream>

#include "../lib/json.h"

using json = nlohmann::json;

void ioMaker(const int board[15][15], bool blackTurn,
             const int output[15][15]);
int randomRotateMirror(int table[15][15]);
void rotateMirror(int table[15][15], int index);
void rotate(int table[15][15]);
void mirror(int table[15][15]);

std::ofstream outfile ("cnndata.txt");

int main() {
  srand((unsigned)time(NULL));

  std::ifstream i("jsondata.json");
  json j;
  i >> j;

  for (auto& game : j) {
    int board[15][15] = {0};
    int count = 0;
    for (auto& play : game["record"]) {
      if (play.is_null()) continue;

      ++count;

      int row = play["r"], col = play["c"];

      // Copy current board.
      int copyBoard[15][15];
      std::copy(&board[0][0], &board[0][0] + 225, &copyBoard[0][0]);

      int output[15][15] = {0};
      output[row][col] = 1;

      int r = randomRotateMirror(copyBoard);
      rotateMirror(output, r);
      ioMaker(copyBoard, (count & 1), output);

      // put chess in board (black: 1, white: -1)
      board[row][col] = (count & 1) ? 1 : -1;
    }
  }

  outfile.close();
}

void ioMaker(const int board[15][15], bool blackTurn,
             const int output[15][15]) {
  int input[4][15][15] = {0};

  // Input feature
  // Black
  for (int r = 0; r < 15; ++r) {
    for (int c = 0; c < 15; ++c) {
      if (board[r][c] == 1) input[0][r][c] = 1;
    }
  }
  // White
  for (int r = 0; r < 15; ++r) {
    for (int c = 0; c < 15; ++c) {
      if (board[r][c] == -1) input[1][r][c] = 1;
    }
  }
  // Empty
  for (int r = 0; r < 15; ++r) {
    for (int c = 0; c < 15; ++c) {
      if (board[r][c] == 0) input[2][r][c] = 1;
    }
  }
  // BlackTurn
  for (int r = 0; r < 15; ++r) {
    for (int c = 0; c < 15; ++c) {
      if (blackTurn) input[3][r][c] = 1;
    }
  }

  static int count = 0;
  if (outfile.is_open()) {
    outfile << ++count << "\n";

    for (auto& feature : input) {
      for (int r = 0; r < 15; ++r) {
        for (int c = 0; c < 15; ++c) {
          outfile << feature[r][c] << " ";
        }
        outfile << "\n";
      }
      outfile << "\n";
    }
    for (int r = 0; r < 15; ++r) {
      for (int c = 0; c < 15; ++c) {
        outfile << output[r][c] << " ";
      }
      outfile << "\n";
    }
    outfile << "\n";
  } else  {
    std::cout << "Unable to open file" << std::endl;
  }
}

int randomRotateMirror(int table[15][15]) {
  int index = rand() % 8;
  for (int m = 0; m < 2; ++m) {
    for (int r = 0; r < 4; ++r) {
      if (index == m * 4 + r) return index;
      rotate(table);
    }
    mirror(table);
  }
}

void rotateMirror(int table[15][15], int index) {
  for (int m = 0; m < 2; ++m) {
    for (int r = 0; r < 4; ++r) {
      if (index == m * 4 + r) return;
      rotate(table);
    }
    mirror(table);
  }
}

void rotate(int table[15][15]) {
  int temp[15][15];
  // rotate 90 degrees clockwise (row -> col, col -> 14 - row)
  for (int r = 0; r < 15; ++r)
    for (int c = 0; c < 15; ++c)
      temp[c][14 - r] = table[r][c];

  for (int r = 0; r < 15; ++r)
    for (int c = 0; c < 15; ++c)
      table[r][c] = temp[r][c];
}

void mirror(int table[15][15]) {
  int temp[15][15];
  // mirror (row -> col, col -> row)
  for (int r = 0; r < 15; ++r)
    for (int c = 0; c < 15; ++c)
      temp[c][r] = table[r][c];

  for (int r = 0; r < 15; ++r)
    for (int c = 0; c < 15; ++c)
      table[r][c] = temp[r][c];
}

File: neuralnetwork/gomoku/jsonreader.py
import json
import random

def outputData(board, blackTurn, output, file, counter):
    # Input feature: self chess, opp chess, empty point.
    input = [[[0] * 15 for _ in range(15)] for _ in range(3)]

    for row in range(15):
        for col in range(15):
            if board[row][col] == 0:  # Empty point.
                input[2][row][col] = 1
            elif board[row][col] == -1:  # Black chess.
                if blackTurn:
                    input[0][row][col] = 1
                else:
                    input[1][row][col] = 1
            else:  # White chess.
                if blackTurn:
                    input[1][row][col] = 1
                else:
                    input[0][row][col] = 1

    file.write(str(counter) + '\n')

    for feature in range(3):
        for row in range(15):
            for col in range(15):
                file.write(str(input[feature][row][col]))
            file.write('\n')
        file.write('\n')

    for row in range(15):
        for col in range(15):
            file.write(str(output[row][col]))
        file.write('\n')
    file.write('\n')

# Rotate and mirror
# There are 8 kinds of results, (4 rotate * 2 mirror),
# key is used to decide result.
def rotateMirror(board, key):
    for m in range(2):
        for r in range(4):
            if key == m * 4 + r:
                return
            rotate(board)
        mirror(board)

def rotate(board):
    temp = [[0] * 15 for _ in range(15)]
    # Rotate 90 degrees clockwise (row -> col, col -> 14 - row).
    for row in range(15):
        for col in range(15):
            temp[col][14 - row] = board[row][col]

    for row in range(15):
        board[row] = temp[row]

def mirror(board):
    temp = [[0] * 15 for _ in range(15)]
    # Mirror (row -> col, col -> row).
    for row in range(15):
        for col in range(15):
            temp[col][row] = board[row][col];

    for row in range(15):
        board[row] = temp[row]

def main():
    # This constant value is the probability of
    # select current board to be one of the training data,
    # it's used to prevent over fitting.
    SELECT_RATE = 0.05  # 5% board will be used

    # Counter
    inputCounter = 0
    outputCounter = 0

    # Read json data.
    file = open('jsondata.json', 'r')
    jsonData = json.loads(file.read())
    file.close()

    outputFile = open('cnndata.txt', 'w')

    for game in jsonData:
        # Check this game exist record.
        if 'record' not in jsonData[game]:
            continue

        board = [[0] * 15 for _ in range(15)]
        count = 0
        for move in jsonData[game]['record']:
            if move == None:
                continue

            inputCounter += 1

            row = move['r']
            col = move['c']

            if random.random() < SELECT_RATE:
                outputCounter += 1

                # Copy current board.
                copyBoard = [row[:] for row in board]
                # Current board's output is next move.
                output = [[0] * 15 for _ in range(15)]
                output[row][col] = 1

                # Random mirror rotate.
                randomMirrorRotateSeed = random.randrange(8)
                rotateMirror(copyBoard, randomMirrorRotateSeed)
                rotateMirror(output, randomMirrorRotateSeed)

                # Output input output data to 'outputFile'
                blackTurn = (count & 1 == 0)
                outputData(copyBoard, blackTurn, output,
                           outputFile, outputCounter)

            # Update board, black: -1, white: 1.
            board[row][col] = 1 if count & 1 == 1 else -1
            count += 1

            # Print input/output info
            print('Input', str(inputCounter),
                  'Output', str(outputCounter), end = '\r')

    outputFile.close()

if __name__ == "__main__":
    main()